### ISSUE TITLE
Could AltertDescription::CertificateRequired be renamed to `ClientCertificateRequired`

### ISSUE THREAD

**Issue by rukai** [rustls/rustls#1307]
I was getting this error and was thinking "duh of course I need a certificate, but I've definitely setup the server with a certificate"
Turns out the server was configured with mutual TLS and the error was referring to the missing client certificate.

If we renamed it from `CertificateRequired` to `ClientCertificateRequired` I think that would help avoid confusions like mine?

I saw this comment:
```rust
    /// The `AlertDescription` TLS protocol enum.  Values in this enum are taken
    /// from the various RFCs covering TLS, and are listed by IANA.
    /// The `Unknown` item is used when processing unrecognised ordinals.
```
But its not clear from the comment if the variant names are taken from the RFC's as well or just the error codes.
I dont know which particular RFCs its talking about so I couldnt check if that was the case.

**Comment by djc** [comment]
The [spec](https://www.rfc-editor.org/rfc/rfc8446#section-6.2) calls it `certificate_required`, so I'm not sure we want to deviate from that in the name. Maybe we could change where/how it is displayed to show more context, but looking at the code I don't think there's an easy way to insert that information today.

**Comment by rukai** [comment]
Fair enough, it would be nice to have more context on those errors but that would be an overhaul much larger than what I've described here so I'll go ahead and close the issue.

**Comment by jsha** [comment]
I agree we should keep the name matching the spec, and also renaming it would be a semver-breaking change, which it would be best to avoid.

However, I think we can make the relevant error message more useful without a breaking change.

@rukai presumably the error you got was a [rustls::Error](https://docs.rs/rustls/latest/rustls/enum.Error.html) with the value `AlertReceived(CertificateRequired)`? Per spec, this is _always_ about a server requesting client certificates:

> certificate_required:  Sent by servers when a client certificate is
> desired but none was provided by the client.

Right now the Display implementation for Error does this:

                Self::AlertReceived(ref alert) => write!(f, "received fatal alert: {:?}", alert),

Which would just display `received fatal alert: CertificateRequired`.

It would be better to say: `received fatal alert: certificate_required: the server desired a client certificate but none was provided by the client`.

To do that, we would change the `{:?}` to `{}` and add a manual `Display` impl on `AlertDescription` that displays the alert name plus an optional user-friendly description.

A slightly larger, but related change: right `AlertDescription` has a derived `Debug` impl provided by `enum_builder!`. That's convenient but slightly wrong, because we print `CertificateRequired` instead of `certificate_required`. The latter is  preferable because it's the actual name in the RFC and registered with IANA. Someone Ctrl-F'ing for CertificateRequired in the TLS RFC would not find that string. We should probably get rid of the derived `Debug` and have `enum_builder!` implement a `Debug` that puts things into lower_snake_case (except for cipher suites, which should stay in SCREAMING_SNAKE_CASE like they are today).

**Comment by rukai** [comment]
Oooh this sounds really nice.

> @rukai presumably the error you got was a [rustls::Error](https://docs.rs/rustls/latest/rustls/enum.Error.html) with the value AlertReceived(CertificateRequired)

Yep that is correct
